File: src/rclcpp/custom_interface_usage/src/sub.cpp
#include "my_interfaces/msg/num.hpp"
#include "my_interfaces/msg/sphere.hpp"
#include "rclcpp/rclcpp.hpp"

using std::placeholders::_1;

using namespace std::chrono_literals;

class MyCustomSub : public rclcpp::Node {
  public:
    MyCustomSub() : Node("my_custom_sub") {
        sub1_ = this->create_subscription<my_interfaces::msg::Num>(
            "my_sphere_topic", 10, std::bind(&MyCustomSub::topic1_callback, this, _1));
        sub2_ = this->create_subscription<my_interfaces::msg::Sphere>(
            "my_sphere_topic", 10, std::bind(&MyCustomSub::topic2_callback, this, _1));
    }

  private:
    void topic1_callback(const my_interfaces::msg::Num::SharedPtr msg) const {
        RCLCPP_INFO(this->get_logger(), "num: %ld", msg->num);
    }
    void topic2_callback(const my_interfaces::msg::Sphere::SharedPtr msg) const {
        RCLCPP_INFO(this->get_logger(), "center.x: %f", msg->center.x);
    }
    rclcpp::Subscription<my_interfaces::msg::Num>::SharedPtr sub1_;
    rclcpp::Subscription<my_interfaces::msg::Sphere>::SharedPtr sub2_;
};

int main(int argc, char* argv[]) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<MyCustomSub>());
    rclcpp::shutdown();
    return 0;
}

File: src/rclpy/py_custom_interface/py_custom_interface/sub.py
import rclpy
from rclpy.node import Node
from my_interfaces.msg import Num, Sphere


class MyCustomSub(Node):
    def __init__(self):
        super().__init__("custom_sub_node")
        self.sub_ = self.create_subscription(Num, "my_int64", self.num_callback, 10)
        self.sub_sphere = self.create_subscription(
            Sphere, "my_sphere", self.sphere_callback, 10
        )

    def num_callback(self, msg):
        self.get_logger().info(f"Received num: {msg.num}")

    def sphere_callback(self, msg):
        self.get_logger().info(
            f"Received sphere: {msg.center.x}, {msg.center.y}, {msg.center.z}, {msg.radius}"
        )


def main():
    rclpy.init()
    my_custom_sub = MyCustomSub()
    rclpy.spin(my_custom_sub)
    rclpy.shutdown()


if __name__ == "__main__":
    main()
